# <<< Made by Cristiano Dos Anjos >>>
cristianodosanjos019@gmail.com

### ...

### THIS IS THE SECUND PART OF A PROJECT OF CHURN MODELING.
### The first part of this project can be accessed by the following link: 

https://github.com/cristianodosanjos/Big.DataScience/blob/main/Bank%20Customers%20Churn%20(1).ipynb

At this part of our project, we will create a machine learning model for deliver who is the customer that want to leave from our bank.

#### About this dataset...

+ We will use a file of a unknow bank, I got this data from kaggle website.

Context: A dataset which contain some customers who are withdrawing their account from the bank due to some loss and other issues with the help this data we try to analyse and maintain accuracy.

Content: What's inside is more than just rows and columns. Make it easy for others to get started by describing how you acquired the data and what time period it represents, too.

#### Let's get started.

#### Importing libraries...

In [32]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics
import statsmodels.formula.api as smf
from sklearn import preprocessing
from sklearn.pipeline import make_pipeline
from sklearn.svm import SVC

#### Finding our dataset...

In [2]:
df = pd.read_csv('/home/tyrslayer/Documents/ipynb.Apo$tila$/UltimaPastaDeProjetos/ChurnFile.csv')
df.head()

Unnamed: 0  RowNumber  CustomerId   Surname  CreditScore Geography  Gender  \
0           0          1    15634602  Hargrave          619    France  Female   
1           1          2    15647311      Hill          608     Spain  Female   
2           2          3    15619304      Onio          502    France  Female   
3           3          4    15701354      Boni          699    France  Female   
4           4          5    15737888  Mitchell          850     Spain  Female   

   Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Exited  For-sum  
0        101348.88       1        1  
1        112542.58       0        1  
2        113931.57       1        1  
3         93826.63       0        1  
4         79084.10       0        1

This dataset was explored and prepared at the previous part. So let's get started the second part.

In [3]:
df.dtypes

Unnamed: 0           int64
RowNumber            int64
CustomerId           int64
Surname             object
CreditScore          int64
Geography           object
Gender              object
Age                  int64
Tenure               int64
Balance            float64
NumOfProducts        int64
HasCrCard            int64
IsActiveMember       int64
EstimatedSalary    float64
Exited               int64
For-sum              int64
dtype: object

In [4]:
X = df[['CreditScore', 'Geography', 'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary']].values
y = df['Exited'].values

Our new variable X will be the params of our model, the variable y will be the predicted response.

We need make some changes. Our model of machine can't perform with columns in format of object, it will be chaged to integers by another model from library preprocessing.

In [5]:
X[:,1][:7]

array(['France', 'Spain', 'France', 'France', 'Spain', 'Spain', 'France'],
      dtype=object)

In [6]:
lbg = preprocessing.LabelEncoder().fit(['France', 'Spain', 'Germany'])
X[:,1] = lbg.transform(X[:,1])

In [7]:
X[:,2]

array(['Female', 'Female', 'Female', ..., 'Female', 'Male', 'Female'],
      dtype=object)

In [8]:
lba = preprocessing.LabelEncoder().fit(['Male', 'Female'])
X[:,2] = lba.transform(X[:,2])

Now let's be using a method from the library "model selection", this method called "train_test_split", will be used to split the data for us get trained and posteriorly test our data. We are going to train 65% of data, and get testing 35%.

In [9]:
X_train, X_test, y_train, y_test= train_test_split(X, y, test_size= 0.35)

Our will be training 65% of data, so this model will have a lot of params for identify parttens in our data.

In [10]:
lr = LogisticRegression()
rfc = RandomForestClassifier(max_depth=2, random_state=0)

In [11]:
lr.fit(X_train, y_train)

LogisticRegression()

In the line above there are our first machine learning model's, called Logistic Regression. Logistic regression is the appropriate regression analysis to conduct when the dependent variable is dichotomous (binary).  Like all regression analyses, the logistic regression is a predictive analysis.  Logistic regression is used to describe data and to explain the relationship between one dependent binary variable and one or more nominal, ordinal, interval or ratio-level independent variables.

In [12]:
rfc.fit(X_train, y_train)

RandomForestClassifier(max_depth=2, random_state=0)

Above we have the Random Forest Classifier. Random Forest is a meta estimator that fits a number of decision tree classifiers on various sub-samples of the dataset and uses averaging to improve the predictive accuracy and control over-fitting. The sub-sample size is controlled with the max_samples parameter if bootstrap=True (default), otherwise the whole dataset is used to build each tree.

Now let's predict and get avaluated the models in our dataset.

In [13]:
lr_ypred = lr.predict(X_test)
rfc_ypred = rfc.predict(X_test)

The r2_score function shows us in percentage how much the resources used to create the model, explain the predicted values. We will use r2_score for bring out that percentage.

In [14]:
metrics.r2_score(y_test, lr_ypred)

-0.273200792121834

In [15]:
metrics.r2_score(y_test, rfc_ypred)

-0.15004623470464384

As we can see, ours models had a very low score. Let's use others functions... 

In [16]:
print(metrics.classification_report(y_test, lr_ypred))

              precision    recall  f1-score   support

           0       0.81      0.98      0.89      2813
           1       0.43      0.07      0.13       687

    accuracy                           0.80      3500
   macro avg       0.62      0.53      0.51      3500
weighted avg       0.74      0.80      0.74      3500



In [17]:
print(metrics.classification_report(y_test, rfc_ypred))

              precision    recall  f1-score   support

           0       0.82      1.00      0.90      2813
           1       1.00      0.08      0.14       687

    accuracy                           0.82      3500
   macro avg       0.91      0.54      0.52      3500
weighted avg       0.85      0.82      0.75      3500



Recall is the model's ability to find all positive instances. The F1 score can be interpreted as a weighted average of the precision and recall, where an F1 score reaches its best value at 1 and worst score at 0. This is all that we need know about the function classification_report.

For recall, let's use other alternative. The method ROC (Receiver Operating Characteristic Curve) that use TPR (True Positive Rate or Recall) and the FPR (False Positive Rate).

In [62]:
print('ROC: %0.2f' % metrics.roc_auc_score(y_test, lr_ypred))

ROC: 0.53


In [61]:
print('ROC: %0.2f' % metrics.roc_auc_score(y_test, rfc_ypred))

ROC: 0.54


We have 53% of ROC in the model Logistic Regression, and 54% in the Random Forest Classifier.

Even here, we can see that the model Random Forest Classifier is a model more appropriate for this dataset than the Logistic Regression model.

Now let's get creating another good classfication's model, this model is called SVM(Support Vector Machines). Let's use a standardization for eliminate some discrepancies in the data, the StandardScaler will help us to converge the data.

In [34]:
msvc = make_pipeline(preprocessing.StandardScaler(), SVC(gamma='auto'))
msvc.fit(X_train, y_train)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('svc', SVC(gamma='auto'))])

In [42]:
msvc_ypred = msvc.predict(X_test)

In [43]:
print(metrics.r2_score(y_test, msvc_ypred))

0.10169616942755388


In [44]:
print(metrics.classification_report(y_test, msvc_ypred))

              precision    recall  f1-score   support

           0       0.87      0.98      0.92      2813
           1       0.79      0.38      0.51       687

    accuracy                           0.86      3500
   macro avg       0.83      0.68      0.71      3500
weighted avg       0.85      0.86      0.84      3500



As we had seen in the previous models, the best r2_score was from Random Forest Classifier with -15%, and as we can see above, the model SVC with standardization have 10% of r2_score and a classifican report better than the previous models. 

Let's create the Random Forest Classifier again, but at this time using standardization too.

In [47]:
mrfc = make_pipeline(preprocessing.StandardScaler(), RandomForestClassifier(max_depth= 2))
mrfc.fit(X_train, y_train)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('randomforestclassifier',
                 RandomForestClassifier(max_depth=2))])

In [51]:
print(metrics.r2_score(y_test, mrfc_ypred))

-0.14823513827203794


In [50]:
mrfc_ypred = mrfc.predict(X_test)
print(metrics.classification_report(y_test, mrfc_ypred))

              precision    recall  f1-score   support

           0       0.82      1.00      0.90      2813
           1       1.00      0.08      0.14       687

    accuracy                           0.82      3500
   macro avg       0.91      0.54      0.52      3500
weighted avg       0.85      0.82      0.75      3500



With the values above, we verified that SVC is definately better than Random Forest for get applying in this dataset.

In [53]:
print(metrics.accuracy_score(y_test, msvc_ypred))

0.8582857142857143


In [55]:
print('AUC: %0.2f' % metrics.roc_auc_score(y_test, msvc_ypred))

AUC: 0.68


In [64]:
print(metrics.f1_score(y_test, msvc_ypred))

0.5108481262327416


#### The name of our model is msvc, it is ready. Some findings about our model:
+ 85.8% of accuracy.
+ 68% of ROC (Receiver Operating Characteristic Curve).
+ 51% of F1-score.
+ 10% R2-score.

## Thanks for working with me.:)